optimize: optimize config compatible module

slievrly opened the pull request:
<!-- Please make sure you have read and understood the contributing guidelines -->

- [ ] I have registered the PR [changes](../changes).

### Ⅰ. Describe what this PR did

optimize config compatible module

### Ⅱ. Does this pull request fix one issue?
<!-- If that, add "fixes #xxx" below in the next line, for example, fixes #97. -->


### Ⅲ. Why don't you add test cases (unit test/integration test)? 


### Ⅳ. Describe how to verify it


### Ⅴ. Special notes for reviews

slievrly committed:
step 1

funky-eyes commented on compatible/src/main/java/io/seata/config/ConfigurationChangeEvent.java:
我认为ConfigurationChangeEvent并不属于用户api,即便用户实现了自己的configuration的实现,但是通过这个configurationchangevent并不能与我们内部的configurationchangelistener互通,这块应该是有损的兼容,需要用户改造适配
I think ConfigurationChangeEvent does not belong to the user api, even if the user realizes his own configuration, but through this configurationchangevent and can not communicate with our internal configurationchangelistener, this piece should be detrimental compatibility, the user needs to adapt

slievrly committed:
step 2

slievrly committed:
step 3

slievrly committed:
optimize: optimize config compatible module

# Conflicts:
#	changes/en-us/2.x.md
#	changes/zh-cn/2.x.md
#	compatible/src/main/java/io/seata/core/context/RootContext.java

slievrly committed:
optimize: optimize config compatible module

# Conflicts:
#	changes/en-us/2.x.md
#	changes/zh-cn/2.x.md
#	compatible/src/main/java/io/seata/core/context/RootContext.java

slievrly committed:
temp

slievrly committed:
add test for config compatible

slievrly committed:
SPI definition use lowercase

slievrly committed:
Merge remote-tracking branch 'upstream/2.x' into 2.x_24_03_09_01

# Conflicts:
#	changes/en-us/2.x.md

funky-eyes commented on config/seata-config-core/src/main/java/io/seata/config/Processor.java:
```diff
@@ -18,9 +18,7 @@
- *
- *
+ * Notes: used for io.seata SPI interface
  */
```
```suggestion
 */
@Deprecated
```

funky-eyes commented on config/seata-config-core/src/main/java/io/seata/config/ConfigurationProvider.java:
```diff
@@ -18,7 +18,13 @@
+ * Notes: used for io.seata SPI interface
  */
-public interface ConfigurationProvider extends org.apache.seata.config.ConfigurationProvider {
+public interface ConfigurationProvider {
```
```suggestion
@Deprecated
public interface ConfigurationProvider {
```

funky-eyes commented on config/seata-config-core/src/main/java/io/seata/config/ConfigurationChangeType.java:
```diff
@@ -16,11 +16,17 @@
- * 
- */
-public interface ExtConfigurationProvider extends org.apache.seata.config.ExtConfigurationProvider {
+public enum ConfigurationChangeType {
```
```suggestion
@Deprecated
public enum ConfigurationChangeType {
```

funky-eyes commented on config/seata-config-core/src/main/java/io/seata/config/ConfigurationChangeListener.java:
```diff
@@ -0,0 +1,91 @@
+
+import org.apache.seata.common.thread.NamedThreadFactory;
+
+public interface ConfigurationChangeListener {
```
```suggestion
@Deprecated
public interface ConfigurationChangeListener {
```

funky-eyes commented on config/seata-config-core/src/main/java/io/seata/config/ConfigurationChangeEvent.java:
```diff
@@ -0,0 +1,138 @@
+ */
+package io.seata.config;
+
+public class ConfigurationChangeEvent {
```
```suggestion
@Deprecated
public class ConfigurationChangeEvent {
```

funky-eyes commented on config/seata-config-core/src/main/java/io/seata/config/Configuration.java:
```diff
@@ -0,0 +1,310 @@
+
+import org.apache.seata.common.util.StringUtils;
+
+public interface Configuration {
```
```suggestion
@Deprecated
public interface Configuration {
```

funky-eyes commented on config/seata-config-core/src/main/java/io/seata/config/AbstractConfiguration.java:
```diff
@@ -0,0 +1,164 @@
+import org.apache.seata.common.util.DurationUtil;
+import org.apache.seata.common.util.StringUtils;
+
+public abstract class AbstractConfiguration implements Configuration {
```
```suggestion
@Deprecated
public abstract class AbstractConfiguration implements Configuration {
```

funky-eyes commented on config/seata-config-core/src/main/java/io/seata/config/FileConfiguration.java:
```diff
@@ -0,0 +1,112 @@
+ * /apache/shardingsphere/transaction/base/seata/at/SeataATShardingSphereTransactionManager.java
+ * 2.EnhancedServiceLoader.load(ExtConfigurationProvider.class).provide(configuration)
+ */
+public class FileConfiguration extends AbstractConfiguration {
```
```suggestion
@Deprecated
public class FileConfiguration extends AbstractConfiguration {
```

slievrly replied:
Based on the conclusions from the previous meeting, Deprecated annotations will be updated uniformly.

funky-eyes commented on config/seata-config-core/src/main/java/org/apache/seata/config/ConfigurationFactory.java:
```diff
@@ -176,6 +178,67 @@ private static Configuration buildConfiguration() {
+        io.seata.config.Configuration oldConfiguration = new io.seata.config.FileConfiguration(configuration);
+        try {
+            io.seata.config.Configuration configurationSPIInstance = EnhancedServiceLoader.load(
+                io.seata.config.ExtConfigurationProvider.class).provide(oldConfiguration);
```
这里似乎没有必要加载，当使用2.1的seata时，不存在有`io.seata.config.ExtConfigurationProvider`的实现
It seems unnecessary to load here. When using Seata 2.1, there is no implementation of `io.seata.config.ExtConfigurationProvider`.

slievrly replied:
fixed

funky-eyes commented on config/seata-config-core/src/main/java/org/apache/seata/config/ConfigurationFactory.java:
```diff
@@ -176,6 +178,67 @@ private static Configuration buildConfiguration() {
+        } catch (Exception exx) {
+            LOGGER.error("failed to load non-spring configuration :{}", exx.getMessage(), exx);
+        }
+        return ORIGIN_FILE_INSTANCE;
```
这里不应该返回ORIGIN_FILE_INSTANCE，更应该直接抛出异常提示，可以比较明确的判断问题
Here, instead of returning ORIGIN_FILE_INSTANCE, it would be more appropriate to directly throw an exception with a clear error message, allowing for a more straightforward identification of the issue.

slievrly replied:
fixed

slievrly replied:
An exception cannot be thrown here because the file mode does not go through SPI. I've broken down the previous logic a bit and extracted the assignment of default values to the outer logic, but overall, it remains consistent with what was there before.

funky-eyes commented on config/seata-config-core/src/main/java/org/apache/seata/config/ConfigurationFactory.java:
```diff
@@ -184,4 +247,107 @@ protected static void reload() {
+
+        @Override
+        public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+            List<String> simpleMethod = Arrays.stream(simpleParamsMethodNames).collect(Collectors.toList());
```
为什么不在提前转换成list，而是在每次invoke的时候转换
Why not convert it into a list beforehand, rather than converting it every time it's invoked?

slievrly replied:
fixed

slievrly committed:
fix review

slievrly committed:
fix review

funky-eyes reviewed (approved):
LGTM
